File: rootToLeaf.java
import java.util.*;
// Approach 1- Iterative DFS (Preorder) [using explicit stack]
class Pair{ 
    Node root;
    int currNumber;
    Pair(Node root,int currNumber){
        this.root=root;
        this.currNumber=currNumber;
    }
}

class Solution{ // TC O(N) SC O(N)
    public int sumNumbers(Node root) {
        if (root==null) return 0;
        Stack<Pair> stack = new Stack<>();
        int sum=0, currNumber=0;
        stack.push(new Pair(root,0));
        while(!stack.isEmpty()){
            Pair p= stack.pop();
            Node currNode=p.root;
            currNumber=p.currNumber; 
            if(currNode!=null){
                currNumber=currNumber*10+currNode.val;
                if(currNode.left==null && currNode.right==null){
                    sum+=currNumber;
                } else {
                    stack.push(new Pair(currNode.left,currNumber));
                    stack.push(new Pair(currNode.right,currNumber));
                } 
            }
        }
        return sum;
    }
}
// Approach 2- Recursive (Preorder) [using call-stack]
class Solution1{ // TC O(N) SC O(N)
    int sum=0;
    public void preorder(Node root, int currNumber){
        if(root!=null){
            currNumber=currNumber*10+root.val;
            if(root.left==null && root.right==null){
                sum+=currNumber;
            }
            preorder(root.left,currNumber);
            preorder(root.right,currNumber);
        }
    }
    public int sumNumbers(Node root) {
        preorder(root,0);
        return sum;
    }
}

class rootToLeaf{
    public static void main(String[] args) {
        BinaryTree bt= new BinaryTree();
        bt.root= new Node(1);
        bt.root.left=new Node(2);
        bt.root.right=new Node(3);
        Solution s= new Solution();
        System.out.println(s.sumNumbers(bt.root));
        Solution1 s1= new Solution1();
        System.out.println(s1.sumNumbers(bt.root));
    }
}

File: rootToLeaf.py
class TreeNode:
	def __init__(self, val=None, left=None, right=None):
		self.val=val 
		self.left=left 
		self.right=right

class Solution:
# Approach 1- Recursive DFS (Preorder) [using call-stack]
	def sumNumbers(self, root): # TC O(n) // SC O(n)
		total=0
		def preorder(root, currNumber):
			nonlocal total 

			if root:
				currNumber=currNumber*10+root.val 

				if not root.left and not root.right:
					total+=currNumber 

				preorder(root.left, currNumber)
				preorder(root.right, currNumber)

			return total 

		return preorder(root,0)

# Approach 2- Iterative DFS (Preorder) [using explicit stack]
	def sumNumbers1(self, root): # TC O(n) // SC O(n)
		total=0 
		currNumber=0
		stack=[(root,0)]
		while stack:
			node, currNumber=stack.pop()
			if node:
				currNumber=currNumber*10+node.val 

				if not node.left and not node.right:
					total+=currNumber
				else:
					stack.append((node.right, currNumber))
					stack.append((node.left, currNumber))

		return total 

root=TreeNode(4)
root.left=TreeNode(9)
root.right=TreeNode(0)
root.left.left=TreeNode(5)
root.left.right=TreeNode(1)
s=Solution()
print(s.sumNumbers(root))
print(s.sumNumbers1(root))